client/core/trade.go

```diff
@@ -1267,7 +1329,89 @@ func (t *trackedTrade) finalizeRedeemAction(match *matchTracker, coinID []byte)
 	return errs.ifAny()
 }
 
+// beginFindRedemption begins a search for the redemption of taker's contract
+// and waits for the search results on a goroutine.
+//
+// This method modifies trackedTrade fields and MUST be called with the
+// trackedTrade mutex lock held for writes.
+func (t *trackedTrade) beginFindRedemption(match *matchTracker) {
+	swapCoinID := dex.Bytes(match.MetaData.Proof.TakerSwap)
+	resultChan, err := t.wallets.fromWallet.FindRedemption(swapCoinID)
+	if err != nil {
+		fromID := t.wallets.fromAsset.ID
+		log.Errorf("failed to start redemption search for match %v, %s contract %s: %v",
+			match.id, unbip(fromID), coinIDString(fromID, swapCoinID), err)
+		return
+	}
+	// Redemption search started, start goroutine to monitor result channel.
+	t.awaitingRedemption[swapCoinID.String()] = match.id
+	go t.waitForRedemption(resultChan)
+}
+
+// waitForRedemption waits on the passed `redemptionChan` for FindRedemption
+// results. Should be run as a goroutine.
+func (t *trackedTrade) waitForRedemption(redemptionChan chan *asset.FindRedemptionResult) {
+	frr := <-redemptionChan
+	if frr == nil {
+		return // channel closed without data
+	}
+
+	fromAsset := t.wallets.fromAsset
+	var match *matchTracker
+
+	t.mtx.Lock()
+	defer t.mtx.Unlock()
+
+	orderID, found := t.awaitingRedemption[frr.ContractCoinID.String()]
+	if found {
+		delete(t.awaitingRedemption, frr.ContractCoinID.String())
+		match, found = t.matches[orderID]
+	}
+	if !found {
+		log.Warnf("received find redemption result for unknown contract coin (%s: %s) on order %s",
+			fromAsset.Symbol, coinIDString(fromAsset.ID, frr.ContractCoinID), t.ID())
+		return
+	}
+
+	if frr.Err != nil {
+		log.Errorf("error finding redemption of taker's %s contract %s for order %s, match %s: %v",
+			fromAsset.Symbol, coinIDString(fromAsset.ID, frr.ContractCoinID), t.ID(), match.id, frr.Err)
+		return
+	}
+
+	if match.Match.Status != order.TakerSwapCast {
+		log.Errorf("received find redemption result at wrong step, order %s, match %s, status %s",
+			t.ID(), match.id, match.Match.Status)
+		return
+	}
+
+	_, _, proof, _ := match.parts()
+	if !t.wallets.toWallet.ValidateSecret(frr.Secret, proof.SecretHash) {
+		log.Errorf("Found invalid redemption of taker's %s contract %s for order %s, match %s: invalid secret %s, hash %s",
+			fromAsset.Symbol, coinIDString(fromAsset.ID, frr.ContractCoinID), t.ID(), match.id, frr.Secret, proof.SecretHash)
+		return
+	}
+
+	// Update the match status and set the secret so that Maker's swap
+	// will be redeemed in the next call to trade.tick().
+	match.SetStatus(order.MakerRedeemed)
+	proof.MakerRedeem = []byte(frr.RedemptionCoinID)
+	proof.Secret = frr.Secret
+	err := t.db.UpdateMatch(&match.MetaMatch)
```

Comment:
For this match to be considered in active, we should set IsRevoked to true, since it may not already be.

For refunds, we have `RefundCoin` set to help detect the dead match to retire and in startup to not load it, but for auto-refunds, it needs to be marked as revoked.  Actually, I think we should mark refunds to revoked too.